Grafana 5.0.2 panics on startup if database URL has special characters in the password

Repro:

```
docker network create grafana

docker run --net=grafana --name postgres -p 5432:5432 -d \
  -e "POSTGRES_USER=grafana" \
  -e "POSTGRES_PASSWORD=sec{ret" \
  postgres:9.6

docker run --net=grafana --name grafana \
  -p 3000:3000 \
  -e "GF_DATABASE_TYPE=postgres" \
  -e "GF_DATABASE_URL=postgres://grafana:sec{ret@postgres:5432/grafana" \
  grafana/grafana:5.0.2
```

Results:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x843a0d]

goroutine 22 [running]:
github.com/grafana/grafana/pkg/setting.applyEnvVariableOverrides()
        /go/src/github.com/grafana/grafana/pkg/setting/setting.go:241 +0x5dd
github.com/grafana/grafana/pkg/setting.loadConfiguration(0xc42003ae68)
        /go/src/github.com/grafana/grafana/pkg/setting/setting.go:415 +0x34b
github.com/grafana/grafana/pkg/setting.NewConfigContext(0xc420028e68, 0x0, 0x0)
        /go/src/github.com/grafana/grafana/pkg/setting/setting.go:473 +0x54
main.(*GrafanaServerImpl).initLogging(0xc4202adb80)
        /go/src/github.com/grafana/grafana/pkg/cmd/grafana-server/server.go:107 +0xba
main.(*GrafanaServerImpl).Start(0xc4202adb80, 0x0, 0x0)
        /go/src/github.com/grafana/grafana/pkg/cmd/grafana-server/server.go:56 +0x58
main.main.func2(0xc4202adb80)
        /go/src/github.com/grafana/grafana/pkg/cmd/grafana-server/main.go:89 +0x2f
created by main.main
        /go/src/github.com/grafana/grafana/pkg/cmd/grafana-server/main.go:87 +0x2c5
```

Not a major bug, there are several workarounds - specifying the database config piecemeal instead of
using an env var, or percent-escaping the special character (in this case as %7B). However, this worked
under 4.5.2, and caused some head-scratching while testing our upgrade to 5.x. Ideally Grafana should
log a "hey, your password should be URL-escaped" error instead of panicking. A mention of this edge
case in http://docs.grafana.org/installation/configuration/#database would also be helpful.

--- a/pkg/cmd/grafana-server/server.go
+++ b/pkg/cmd/grafana-server/server.go
@@ -111,7 +111,7 @@ func (g *GrafanaServerImpl) initLogging() {
 	})
 
 	if err != nil {
-		g.log.Error(err.Error())
+		fmt.Fprintf(os.Stderr, "Failed to start grafana. error: %s\n", err.Error())
 		os.Exit(1)
 	}
 

--- a/pkg/setting/setting.go
+++ b/pkg/setting/setting.go
@@ -223,7 +223,7 @@ func shouldRedactURLKey(s string) bool {
 	return strings.Contains(uppercased, "DATABASE_URL")
 }
 
-func applyEnvVariableOverrides() {
+func applyEnvVariableOverrides() error {
 	appliedEnvOverrides = make([]string, 0)
 	for _, section := range Cfg.Sections() {
 		for _, key := range section.Keys() {
@@ -238,7 +238,10 @@ func applyEnvVariableOverrides() {
 					envValue = "*********"
 				}
 				if shouldRedactURLKey(envKey) {
-					u, _ := url.Parse(envValue)
+					u, err := url.Parse(envValue)
+					if err != nil {
+						return fmt.Errorf("could not parse environment variable. key: %s, value: %s. error: %v", envKey, envValue, err)
+					}
 					ui := u.User
 					if ui != nil {
 						_, exists := ui.Password()
@@ -252,6 +255,8 @@ func applyEnvVariableOverrides() {
 			}
 		}
 	}
+
+	return nil
 }
 
 func applyCommandLineDefaultProperties(props map[string]string) {
@@ -377,7 +382,7 @@ func loadSpecifedConfigFile(configFile string) error {
 	return nil
 }
 
-func loadConfiguration(args *CommandLineArgs) {
+func loadConfiguration(args *CommandLineArgs) error {
 	var err error
 
 	// load config defaults
@@ -395,7 +400,7 @@ func loadConfiguration(args *CommandLineArgs) {
 	if err != nil {
 		fmt.Println(fmt.Sprintf("Failed to parse defaults.ini, %v", err))
 		os.Exit(1)
-		return
+		return err
 	}
 
 	Cfg.BlockMode = false
@@ -413,7 +418,10 @@ func loadConfiguration(args *CommandLineArgs) {
 	}
 
 	// apply environment overrides
-	applyEnvVariableOverrides()
+	err = applyEnvVariableOverrides()
+	if err != nil {
+		return err
+	}
 
 	// apply command line overrides
 	applyCommandLineProperties(commandLineProps)
@@ -424,6 +432,8 @@ func loadConfiguration(args *CommandLineArgs) {
 	// update data path and logging config
 	DataPath = makeAbsolute(Cfg.Section("paths").Key("data").String(), HomePath)
 	initLogging()
+
+	return err
 }
 
 func pathExists(path string) bool {
@@ -471,7 +481,10 @@ func validateStaticRootPath() error {
 
 func NewConfigContext(args *CommandLineArgs) error {
 	setHomePath(args)
-	loadConfiguration(args)
+	err := loadConfiguration(args)
+	if err != nil {
+		return err
+	}
 
 	Env = Cfg.Section("").Key("app_mode").MustString("development")
 	InstanceName = Cfg.Section("").Key("instance_name").MustString("unknown_instance_name")

--- a/pkg/setting/setting_test.go
+++ b/pkg/setting/setting_test.go
@@ -37,6 +37,13 @@ func TestLoadingSettings(t *testing.T) {
 			So(appliedEnvOverrides, ShouldContain, "GF_SECURITY_ADMIN_PASSWORD=*********")
 		})
 
+		Convey("Should replace password when defined in environment2", func() {
+			os.Setenv("GF_DATABASE_URL", "postgres://grafana:sec{ret@postgres:5432/grafana")
+			err := NewConfigContext(&CommandLineArgs{HomePath: "../../"})
+
+			So(err, ShouldNotBeNil)
+		})
+
 		Convey("Should replace password in URL when url environment is defined", func() {
 			os.Setenv("GF_DATABASE_URL", "mysql://user:secret@localhost:3306/database")
 			NewConfigContext(&CommandLineArgs{HomePath: "../../"})

